### ISSUE TITLE
Google Apps SAML Response Invalid

### ISSUE THREAD

**Issue by atticus88** [apereo/cas#2413]
## Issue

CAS Version: 5.0.3
Server: Tomcat 8

Google Apps Integration shows the following error when logging in. `G Suite - This account cannot be accessed because we could not parse the login request.`

I think I have been able to trace it back to the `InResponseTo=""` attribute value in the SAML response. According to the [saml-schema-protocol-2.0](https://docs.oasis-open.org/security/saml/v2.0/saml-schema-protocol-2.0.xsd) specification the `InResponseTo` attribute value should be `type="NCName"` (non-colonized name) which cannot contain several symbol characters ` `, `!`, `"`, `#`, `$`, `%`, `&`, `'`, `(`, `)`, `*`, `+`, `,`, `/`, `:`, `;`, `<`, `=`, `>`, `?`, `@`, `[`, `\`, `]`, `^`,  `{`, `|`, `}`, `~`. The value of the field is being set to the serviceId which is a URL and contains a `://`, making the response invalid.

Failing Response:
```xml
<samlp:Response 
        xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol"
        xmlns="urn:oasis:names:tc:SAML:2.0:assertion" 
        xmlns:xenc="http://www.w3.org/2001/04/xmlenc#" 
	ID="opflgcfnepkmjegiofpmnhcgfjiepddfanbdoham" 
	InResponseTo="https://www.google.com/a/my.domain.com/acs" 
	IssueInstant="2017-02-23T18:09:11.279Z" Version="2.0">
```

**Comment by mmoayyed** [comment]
So just confirming, making the InResponseTo be a valid NCName did actually resolve this issue and you were able to get to the G Suite ?

**Comment by atticus88** [comment]
I have not been able to build and run a version with the change included, but I did verify with the following tool that it would generate valid response by removing the restricted characters: https://www.samltool.com/validate_response.php

**Comment by mmoayyed** [comment]
OK, great. If you can provide further confirmation that the change does in fact fix the issue with GApps, that would be much better. You want to make the change to the relevant modules, `build` and `install` them in your local repo, then in your overlay switch the CAS version to the newly built instance and try again.

**Comment by mmoayyed** [comment]
Excellent. Could you turn that change into a PR please?

**Comment by atticus88** [comment]
Done.

**Comment by mmoayyed** [comment]
@atticus88 Given this is a very small change, do you mind if I add it directly myself? Your form seems to be way too outdates, and as a result the PR has a lot of left-over files. If you don't feel like merging and rebasing, I'd be happy to merge this "on your behalf"

LMK.

**Comment by mmoayyed** [comment]
And having it reviewed, I think the there is a bug in the way responses are built. See:

```java
public Response newResponse(final String id, final ZonedDateTime issueInstant,
                                final String recipient, final WebApplicationService service) {

        final Response samlResponse = newSamlObject(Response.class);
        samlResponse.setID(id);
        samlResponse.setIssueInstant(DateTimeUtils.dateTimeOf(issueInstant));
        samlResponse.setVersion(SAMLVersion.VERSION_20);
        samlResponse.setInResponseTo(recipient); //this is the problem. Remove this line maybe? 
        setInResponseToForSamlResponseIfNeeded(service, samlResponse);
        return samlResponse;
}
```

**Comment by atticus88** [comment]
I am fine with you making the change.

**Comment by mmoayyed** [comment]
Cool. Please do try the new snapshot as soon as this is passed https://travis-ci.org/apereo/cas/builds/205939941

**Comment by atticus88** [comment]
What is the name of the snapshot?

**Comment by mmoayyed** [comment]
https://github.com/apereo/cas/blob/5.0.x/gradle.properties#L20

**Comment by atticus88** [comment]
That worked! Thanks!

**Comment by tfrancois** [comment]
Hi  mmoayyed, atticus88.

I confirmed that the bug fix has not been incorporated in 5.1.0.RC3-SNAPSHOT.  As a result, the issue has returned.  Can we make the recommended change posted above and incorporate into 5.1.0.RC3+ future builds?  I am not famaliar on how to compile from source and do the patch.

Please advise and thank you for all your continued hard work and support.

**Comment by mmoayyed** [comment]
Follow this guide:
https://apereo.github.io/cas/developer/Build-Process.html

And this guide:
https://apereo.github.io/cas/developer/Contributor-Guidelines.html

**Comment by mmoayyed** [comment]
I also suggest you try with 5.1.0.RC4-SNAPSHOT.

**Comment by tfrancois** [comment]
Hi, mmoayyed.

> I also suggest you try with 5.1.0.RC4-SNAPSHOT.

I tried 5.1.0.RC4-SNAPSHOT and it did not solve the issue.  I also confirmed it was not a problem with my x509 cert as some others have suggested because I used the same cert and private key with a legacy CAS 3.5.2 build and it worked perfectly.

Any way that the change/fix that closed this topic get merged into the RC4-snapshot build?  I've tried everything else to no avail other than building and compiling my own build which I am terrified of doing.  I'd rather leave that to the experts/more experienced devs.  Please help.

Thanks again.

**Comment by mmoayyed** [comment]
I can try to take a pass, but there is no telling on when, if, and how I might get to this. Not working on a deployment that needs this, so not exactly a priority issue for me. 

If you have a Java background, then you can surely follow the guides I pasted above to get the build working, and simply apply the patch posted here, test it and then share the change as a pull request. It's a process you need to learn and get comfortable with. Start now and I can help where needed. That would be the fastest way to get this resolved.

**Comment by tfrancois** [comment]
I definitely gave it a go but gave up after a few hours (and I usually don't on these types of things).  

Should we reopen #2414 as a result? Releases 5.1+ are way too good for me to even consider rolling back to an earlier version and I didn't even realize how different they were until this issue forced me to take a look.  Could you give a timetable on when the patch could be in? 

Thanks.

**Comment by mmoayyed** [comment]
Outstanding work. Thank you for your perseverance. I'll be happy to apply the patch manually.

>For whatever reason, I couldn't compile the FULL build

I would like to further assist you make progress here. Getting the build to fully compile is the first step in making sure you can continue to contribute to the project. It's *more important* than any other issue. 

What seems to be the problem? 

>I have never committed changes back up to Github and fearful that I won't do it properly 

No such thing! Shed your fears. There is nothing wrong with you (or anyone) making a mistake. The only way to learn it is to practice and make mistakes. Be not afraid. All attempts and efforts are more than welcome, and we are here to ensure you can make progress independently and successfully. We'll meet you half way, and with practice you can make good progress and we'd just merge :) 

This guide should explain the basic idea:
https://apereo.github.io/cas/developer/Contributor-Guidelines.html

In short, 

- you create a branch off of the branch you want to use. `master` in this case. Name your branch anything you like.
- make changes, commit and push to your own branch of your own fork.
- go to github, and create a pull request from your fork and branch to CAS.
- That's it.

Or put another way:

```
- Fork the apereo/CAS project
- git clone <your own fork>
- git checkout master # the branch you want your changes to apply
- git checkout -b the-issue-that-bothers-me-branch
- # make changes, build and test
- git commit -A .
- git push origin the-issue-that-bothers-me-branch
- Go to github and create a pull request from your fork/the-issue-that-bothers-me-branch to CAS/<master> or whatever branch you want your changes to be pushed into.
```

**Comment by mmoayyed** [comment]
Also, I had a look at your patch. It only includes jar files. Those wont work. You need to pass along the changes you actually made to the source code, so they can be applied here. We can not ship "your version" of some JAR with CAS.

Thanks again.

**Comment by tfrancois** [comment]
Thanks @mmoayyed.  I am so glad to be able to help and contribute to this outstanding project.  In my opinion, the 5.1 release is the best CAS version to date.

Attached (including full path from cas-server build folder root path):
[cas-server-support-saml-googleapps-core-5.1.0-RC4-SNAPSHOT-patch.zip](https://github.com/apereo/cas/files/951714/cas-server-support-saml-googleapps-core-5.1.0-RC4-SNAPSHOT-patch.zip)

**Comment by mmoayyed** [comment]
You're most welcome. I appreciate you sticking in with the release candidates and testing. I'll review and incorporate your patch later today, if the stars align.
